File: src/algo/transfer.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   transfer.c                                         :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: vpolard <vpolard@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2026/02/28 09:46:44 by vpolard           #+#    #+#             */
/*   Updated: 2026/03/08 19:02:12 by vpolard          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "algo.h"

static void    get_closest(t_package *package)
{
	int	closest;
	int	min_diff;
	int	current_diff;
	int	index;

	closest = 0;
	min_diff = ft_abs(package->b->content[closest]
		- package->a->content[package->a->pin]);
	index = 0;
	while (index <= package->b->btm)
	{
		current_diff = ft_abs(package->b->content[index]
			- package->a->content[package->a->pin]);
		if (current_diff < min_diff)
		{
			min_diff = current_diff;
			closest = index;
		}
		index++;
	}
	package->b->pin = closest;
}

static void    get_directions(t_package *package)
{
	t_move *move;

	move = package->current;
	move->apin = package->a->pin;
	move->bpin = package->b->pin;
	move->aval = package->a->content[move->apin];
	move->bval = package->b->content[move->bpin];
	move->abtm = package->a->btm;
	move->bbtm = package->b->btm;
	move->adir = 1;
	if (move->apin > (move->abtm / 2))
		move->adir = 0;
	move->bdir = 1;
	if (move->bpin > (move->bbtm / 2))
		move->bdir = 0;
}

static void	get_individual_cost(t_package *package)
{
	t_move *move;

	move = package->current;
	if (move->adir)
		move->acost = move->apin;
	if (!move->adir)
		move->acost = move->abtm - move->apin + 1; 
	if (move->bdir)
		move->bcost = move->bpin;
	if (!move->bdir)
		move->bcost = move->bbtm - move->bpin + 1;
	if (move->aval < move->bval)
		move->bcost++;
	if (move->abtm == 0)
		move->acost = 0;
	if (move->bbtm == 0)
		move->bcost = 0;
}

static void	get_compared_cost(t_package *package)
{
	t_move *move;

	move = package->current;
	if (move->adir == move->bdir)
		move->cost = ft_biggest(move->acost, move->bcost);
	if (is_compatible(package->current))
	{
		if (move->acost > move->bcost)
			move->cost = move->acost;
		if (move->acost < move->bcost)
			move->cost = move->bcost;
	}
	if (!move->cost && (move->acost || move->bcost))
		move->cost = move->acost + move->bcost;
	move->cost++;
}

void    transfer(t_package *package)
{
	package->a->pin = package->a->btm;
	get_closest(package);
	printf("[i] %d (in a) is attached to %d (in b)\n",
		package->a->content[package->a->pin],
		package->b->content[package->b->pin]);
	get_directions(package);
	printf("[i] adir (%d) bdir (%d)\n",
		package->current->adir,
		package->current->bdir);
	get_individual_cost(package);
	printf("[i] acost (%d) bcost (%d)\n",
		package->current->acost,
		package->current->bcost);
	get_compared_cost(package);
	printf("[i] cost (%d)\n", package->current->cost);
	/*
	while (package->a->btm > 2)
	{
		package->a->pin = package->a->top;
		while (package->a->pin >= 0)
		{
			get_closest(package);
			get_directions(package);
			get_individual_cost(package);
			get_compared_cost(package);
			compare_moves(package);
			package->a->pin--;
		}
		apply_move(package);
	}
	*/
}

File: rust/src/algo/transfer.rs
use super::Package;

impl Package {
    pub fn transfer(&mut self) -> &mut Self {
        let mut index: usize;
        while self.a.content.len() > 3 {
            index = 0;
            while index < self.a.content.len() {
                self.get_closest(index)
                    .get_directions()
                    .get_individual_cost()
                    .get_compared_cost()
                    .compare_moves()
                    .apply_move();
                index += 1;
            }
        }
        self
    }

    fn get_closest(&mut self, index: usize) -> &mut Self {
        let mut min_diff: i32 = self.a.content[index] - self.b.content[index];
        self.current.apin = index;
        self.current.aval = self.a.content[index];
        self.current.bpin = 0;
        self.current.bval = self.b.content[0];
        for (pin, element) in self.b.content.iter().enumerate() {
            if (self.a.content[index] - element).abs() < min_diff {
                min_diff = (self.a.content[index] - element).abs();
                self.current.bval = *element;
                self.current.bpin = pin;
            }
        }
        self
    }
    fn get_directions(&mut self) -> &mut Self {
        self.current.adir = true;
        if self.current.apin > self.a.content.len() / 2 {
            self.current.adir = false;
        }
        self.current.bdir = false;
        if self.current.bpin > self.b.content.len() / 2 {
            self.current.bdir = false;
        }
        self
    }
    fn get_individual_cost(&mut self) -> &mut Self {
        match self.current.adir {
            true => self.current.acost = self.current.apin,
            false => self.current.acost = self.a.content.len() - self.current.apin + 1,
        }
        match self.current.bdir {
            true => self.current.bcost = self.current.bpin,
            false => self.current.bcost = self.b.content.len() - self.current.bpin + 1,
        }
        self
    }
    fn get_compared_cost(&mut self) -> &mut Self {
        self
    }
    fn compare_moves(&mut self) -> &mut Self {
        self
    }
    fn apply_move(&mut self) {}
}
